src/radiation_system.hpp

```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 
-		if constexpr (gamma_ != 1.0) {
+		for (int g = 0; g < nGroups_; ++g) {
+			EradVec_guess[g] = NAN;
+		}
+
+		// if constexpr (gamma_ != 1.0) {
+		if (gamma_ != 1.0) {
 			Egas0 = ComputeEintFromEgas(rho, x1GasMom0, x2GasMom0, x3GasMom0, Egastot0);
 			Ekin0 = Egastot0 - Egas0;
 
-			AMREX_ASSERT(Src >= 0.0);
+			AMREX_ASSERT(min(Src) >= 0.0);
 			AMREX_ASSERT(Egas0 > 0.0);
-			AMREX_ASSERT(Erad0 > 0.0);
 
-			const double Etot0 = Egas0 + (c / chat) * (Erad0 + Src);
+			const double Etot0 = Egas0 + (c / chat) * (Erad0 + sum(Src));
 
 			// BEGIN NEWTON-RAPHSON LOOP
 			Egas_guess = Egas0;
-			Erad_guess = Erad0;
+			EradVec_guess = Erad0Vec;
 
 			const double a_rad = radiation_constant_;
 			double F_G = NAN;
-			double F_R = NAN;
-			double rhs = NAN;
-			double kappa = NAN;
 			double fourPiB = NAN;
-			double drhs_dEgas = NAN;
 			double dFG_dEgas = NAN;
-			double dFG_dErad = NAN;
-			double dFR_dEgas = NAN;
-			double dFR_dErad = NAN;
-			double eta = NAN;
-			double deltaErad = NAN;
 			double deltaEgas = NAN;
+			double Rtot = NAN;
+			double dRtot_dEgas = NAN;
+			quokka::valarray<double, nGroups_> Rvec;
+			quokka::valarray<double, nGroups_> dRtot_dErad;
+			quokka::valarray<double, nGroups_> dRvec_dEgas;
+			quokka::valarray<double, nGroups_> dFG_dErad;
+			quokka::valarray<double, nGroups_> dFR_dEgas;
+			quokka::valarray<double, nGroups_> dFR_i_dErad_i;
+			quokka::valarray<double, nGroups_> deltaErad;
+			quokka::valarray<double, nGroups_> F_R;
 
 			const double resid_tol = 1.0e-10; // 1.0e-15;
 			const int maxIter = 400;
 			int n = 0;
-			for (n = 0; n < maxIter; ++n) {
-
+			for (; n < maxIter; ++n) {
 				// compute material temperature
 				T_gas = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Egas_guess, massScalars);
 				AMREX_ASSERT(T_gas >= 0.);
 
 				// compute opacity, emissivity
-				kappa = RadSystem<problem_t>::ComputePlanckOpacity(rho, T_gas);
-				AMREX_ASSERT(kappa >= 0.);
 				fourPiB = chat * a_rad * std::pow(T_gas, 4);
+				auto B = fourPiB / (4.0 * pi);
+				kappaPVec = ComputePlanckOpacity(rho, T_gas);
+				kappaFVec = ComputeFluxMeanOpacity(rho, T_gas);
+				kappaEVec = kappaFVec; // TODO(CCH): later, define ComputeEnergyMeanOpacity
 
 				// compute derivatives w/r/t T_gas
-				const double dB_dTgas = (4.0 * fourPiB) / T_gas;
-				const double dkappa_dTgas = RadSystem<problem_t>::ComputePlanckOpacityTempDerivative(rho, T_gas);
+				const double dB_dTgas = (4.0 * B) / T_gas;
 
 				// compute residuals
-				rhs = dt * (rho * kappa) * (fourPiB - chat * Erad_guess);
-				F_G = (Egas_guess - Egas0) + ((c / chat) * rhs);
-				F_R = (Erad_guess - Erad0) - (rhs + Src);
+				quokka::valarray<amrex::Real, nGroups_> radEnergyFractions{};
+				// TODO(CCH): fix this: if constexpr (nGroups_ > 1)
+				// if constexpr (nGroups_ > 1) {
+				// ../src/radiation_system.hpp(1089): error: An extended __device__ lambda cannot first-capture variable in constexpr-if context
+				// detected during: ...
```

Comment:
This is an annoying limitation of the CUDA compiler. You can use this variable before the `ParallelFor` and then it will work.

Reply:
Good. Actually I realised this before but forgot to put the constexpr back.